File: androidApp/src/main/kotlin/com/virginactive/android/ui/home/HomeViewModel.kt
package com.virginactive.android.ui.home

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.virginactive.shared.domain.error.AppResult
import com.virginactive.shared.domain.error.DomainError
import com.virginactive.shared.domain.home.GetHomeManifestUseCase
import com.virginactive.shared.domain.home.HomeManifest
import com.virginactive.shared.domain.profile.GetProfileUseCase
import kotlinx.coroutines.launch

class HomeViewModel(
    private val getManifest: GetHomeManifestUseCase,
    private val getProfile: GetProfileUseCase,
) : ViewModel() {

    var state by mutableStateOf(HomeUiState())
        private set

    fun load() {
        if (state.isLoading) return
        state = state.copy(isLoading = true, error = null)
        viewModelScope.launch {
            val clubId = when (val profile = getProfile()) {
                is AppResult.Ok -> profile.value.homeClub.id
                is AppResult.Err -> ""
            }
            when (val result = getManifest(clubId)) {
                is AppResult.Ok -> state = state.copy(isLoading = false, manifest = result.value)
                is AppResult.Err -> state = state.copy(isLoading = false, error = result.error)
            }
        }
    }
}

data class HomeUiState(
    val isLoading: Boolean = false,
    val error: DomainError? = null,
    val manifest: HomeManifest? = null,
)

File: iosApp/iosApp/iosApp/Home/HomeViewModel.swift
import Foundation
import Combine
import Shared

@MainActor
final class HomeViewModel: ObservableObject {
    @Published var isLoading: Bool = false
    @Published var error: DomainError?
    @Published var manifest: HomeManifest?

    private let getProfileUseCase: GetProfileUseCase
    private let getHomeManifestUseCase: GetHomeManifestUseCase

    private var loadTask: Task<Void, Never>?

    init(component: AppComponent) {
        self.getProfileUseCase = component.getProfileUseCase
        self.getHomeManifestUseCase = component.getHomeManifestUseCase
    }

    func load() {
        guard !isLoading else { return }
        isLoading = true
        error = nil
        loadTask = Task {
            let clubId: String
            switch onEnum(of: try? await getProfileUseCase.invoke()) {
            case .ok(let ok)?:
                clubId = (ok.value as? UserProfile)?.homeClub.id ?? ""
            default:
                clubId = ""
            }

            guard !Task.isCancelled else { return }

            switch onEnum(of: try? await getHomeManifestUseCase.invoke(clubId: clubId)) {
            case .ok(let ok)?:
                manifest = ok.value as? HomeManifest
                isLoading = false
            case .err(let err)?:
                error = err.error
                isLoading = false
            case .none:
                isLoading = false
            }
        }
    }

    func cancel() {
        loadTask?.cancel()
    }
}
